### PR TITLE
[BH-1792] Updated button handling during pre wake up

### PR DESCRIPTION
**Pull request by ghost**
Button action behavior has been updated for pre-wake. Now, if pre-wake is in progress, the first press, in addition to pressing the front light, disables the pre-wake function.

**Your checklist for this pull request**

Make sure that this PR:
- [x] Complies with our guidelines for contributions
- [ ] Has unit tests if possible
- [ ] Has documentation updated
- [x] Has changelog entry added

### CODE REVIEW

File: module-services/service-time/AlarmServiceAPI.cpp [hunk 1]
```diff
@@ -74,6 +73,26 @@ namespace alarms
...
+
+        bool requestGetPreWakeUpStatus(sys::Service *serv, bool &response)
+        {
+            auto msg = std::make_shared<GetPreWakeUpRequestMessage>();
+            auto ret = serv->bus.sendUnicastSync(msg, service::name::service_time, 1000);
+
+            if (ret.first == sys::ReturnCodes::Success) {
+                auto getPreWakeUpResponse = std::dynamic_pointer_cast<GetPreWakeUpResponseMessage>(ret.second);
+                if ((getPreWakeUpResponse != nullptr) && (getPreWakeUpResponse->retCode == sys::ReturnCodes::Success)) {
+                    LOG_DEBUG("Is Pre-wake up active: %s", getPreWakeUpResponse->active ? "yes" : "no");
```
Comment (patrycja-paczkowska) [inline module-services/service-time/AlarmServiceAPI.cpp:85]: this entry generates a lot of spam in the logs regardless of whether the alarm is set or not

File: products/BellHybrid/services/time/AlarmOperations.cpp [hunk 2]
```diff
@@ -347,6 +347,23 @@ namespace alarms
         AlarmOperationsCommon::handleAlarmEvent(event, alarmType, newStateOn);
     }
 
+    void AlarmOperations::turnOffPreWakeUp(IAlarmOperations::OnTurnOffPreWakeUp callback)
+    {
+        auto nextEvent = getNextPreWakeUpEvent();
+        if (nextEvent.isValid()) {
+            if (auto event = std::dynamic_pointer_cast<AlarmEventRecord>(nextEvent.parent); event) {
+                disablePreWakeUp(event);
```
Comment (patrycja-paczkowska) [inline products/BellHybrid/services/time/AlarmOperations.cpp:355]: Doesn't work in the case where pre-wake up chime and pre-wake up frontlight have different time settings. From what I understand this function only affects active events (pre-wake up light or chime in progress).

**Review by dawidwojtas-mudita — APPROVED** [review]

File: products/BellHybrid/services/audio/ServiceAudio.cpp [hunk 3]
```diff
@@ -197,18 +196,23 @@ namespace service
...
 
+        // if stopType is not provided then stop all inputs, otherwise stop specific ones from stopType
         for (auto &input : audioMux.GetAllInputs()) {
-            auto t = input.token;
-            if (token.IsValid() && t == token) {
-                retCodes.emplace_back(t, stopInput(&input));
-            }
-            if (token.IsUninitialized()) {
-                retCodes.emplace_back(t, stopInput(&input));
+            const auto &currentOperation = input.audio->GetCurrentOperation();
```
Comment (Lefucjusz) [inline products/BellHybrid/services/audio/ServiceAudio.cpp:206]: I don't really understand why this change is needed, what case requires stopping of the specified input?

File: products/BellHybrid/services/audio/ServiceAudio.cpp [hunk 4]
```diff
@@ -197,18 +196,23 @@ namespace service
...
+            if (stopTypes.empty() ||
+                std::find(stopTypes.begin(), stopTypes.end(), currentOperation.GetPlaybackType()) != stopTypes.end()) {
```
Comment (Lefucjusz) [inline products/BellHybrid/services/audio/ServiceAudio.cpp:208]: I'd move it out to some temporary variable to avoid such convoluted if statement:
```suggestion
                const auto isOperationInStopTypes = std::find(stopTypes.begin(), stopTypes.end(), currentOperation.GetPlaybackType()) != stopTypes.end());
                if (stopTypes.empty() || isOperationInStopTypes) {
```

**Review by Lefucjusz — APPROVED** [review]

File: products/BellHybrid/apps/Application.cpp [hunk 5]
```diff
@@ -134,7 +142,28 @@ namespace app
...
         onKeyPressed();
-        return ApplicationCommon::handleKBDKeyEvent(msgl);
+        if (this->getState() != app::ApplicationCommon::State::ACTIVE_FORGROUND) {
+            LOG_FATAL("Terrible terrible damage! Application with no focus grabbed key!");
+        }
+        const auto msg        = static_cast<sevm::KbdMessage *>(msgl);
+        const auto inputEvent = keyTranslator->translate(msg->key);
+        const auto key        = mapKey(inputEvent.getKeyCode());
+        if (!inputEvent.is(gui::KeyCode::KEY_UNDEFINED) && inputEvent.isShortRelease() && key != KeyMap::Frontlight &&
+            key != KeyMap::DeepPressDown && key != KeyMap::DeepPressUp && alarmModel->isPreWakeUpActive()) {
```
Comment (Maciej-Mudita) [inline products/BellHybrid/apps/Application.cpp:152]: If I see correctly, the synchro message is sent here? IMO it shouldn't be like that :/ please don't merge this, I want to take a closer look at it

**Review by Maciej-Mudita — APPROVED** [review]

File: products/BellHybrid/services/audio/ServiceAudio.cpp (hunk 3 above)
Comment (Lefucjusz) [reply products/BellHybrid/services/audio/ServiceAudio.cpp:207]: I get it, thanks for the explanation!

File: products/BellHybrid/services/time/tests/test-BellAlarmOperations.cpp [hunk 6]
```diff
@@ -105,14 +105,16 @@ namespace
         std::unique_ptr<alarms::PreWakeUpSettingsProvider> &&preWakeUpSettingsProvider,
         std::unique_ptr<alarms::SnoozeChimeSettingsProvider> &&snoozeChimeSettingsProvider,
         std::unique_ptr<alarms::OnboardingSettingsProvider> &&onboardingSettingsProvider,
-        std::unique_ptr<alarms::AbstractBedtimeSettingsProvider> &&bedtimeSettingsProvider)
+        std::unique_ptr<alarms::AbstractBedtimeSettingsProvider> &&bedtimeSettingsProvider,
+        alarms::PreWakeUp::ChangeStateCallback callback = [](bool) {})
```
Comment (Lefucjusz) [inline products/BellHybrid/services/time/tests/test-BellAlarmOperations.cpp:109]: You can just default this callback to `nullptr`, as far as I can see there's a check for null pointer before the function is called in AlarmOperations
